### PR TITLE
feat: add feature to record multi-host system

### PR DESCRIPTION
**Pull request by ymski**
## Description

This PR adds feature to record multi-host system.

## Related links

- https://github.com/tier4/ros2caret/pull/163
- https://github.com/tier4/caret_trace/pull/262
- https://github.com/tier4/caret_doc/pull/190

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines](https://github.com/tier4/caret/blob/main/.github/CONTRIBUTING.md).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] (Optional) The PR has been properly tested with CARET_report verification.
- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: src/caret_analyze/plot/timeseries/frequency_timeseries.py [hunk 1]
```diff
@@ -139,19 +139,14 @@ def row_filter_communication(record) -> bool:
...
-                first_timestamps.append(first_timestamp)
-            last_timestamp = records.get_column_series(records.columns[0])[-1]
-            if isinstance(last_timestamp, int):
-                last_timestamps.append(last_timestamp)
-
-        if len(first_timestamps) == 0 or len(last_timestamps) == 0:
+            timestamps = records.get_column_series(records.columns[0])
+            all_timestamps.extend(Util.filter_items(lambda x, : isinstance(x, int), timestamps))
+
+        if len(all_timestamps) == 0 or len(all_timestamps) == 0:
```
Comment (isp-uetsuki) [inline src/caret_analyze/plot/timeseries/frequency_timeseries.py:149]: Minor fix of duplicated code.

```suggestion
        if len(all_timestamps) == 0:
```

File: src/caret_analyze/record/records_service/latency.py [hunk 2]
```diff
@@ -81,15 +83,21 @@ def to_records(
         records = self._create_empty_records()
 
         for start_ts, end_ts in zip(self._start_timestamps, self._end_timestamps):
+            latency = end_ts - start_ts
+            if latency < 0:
+                warn('Record data is invalid. '
+                     'The end time of the path is recorded before the start time.',
```
Comment (isp-uetsuki) [inline src/caret_analyze/record/records_service/latency.py:89]: Latency is for callback or communication, not for path. So I suggest changing the message.

```suggestion
                     'The end time of the callback or communication is recorded before the start time.',
```

File: src/caret_analyze/record/records_service/latency.py [hunk 3]
```diff
@@ -81,15 +83,21 @@ def to_records(
...
+                     UserWarning)
+                continue
             if converter:
                 record = {
                     self._start_column: round(converter.convert(start_ts)),
-                    'latency': end_ts - start_ts
+                    'latency': max(end_ts - start_ts, 0)
```
Comment (isp-uetsuki) [inline src/caret_analyze/record/records_service/latency.py:95]: I suggest using variable defined above.

```suggestion
                    'latency': latency
```

File: src/caret_analyze/record/records_service/latency.py [hunk 4]
```diff
@@ -81,15 +83,21 @@ def to_records(
...
                 }
             else:
                 record = {
                     self._start_column: start_ts,
-                    'latency': end_ts - start_ts
+                    'latency': max(end_ts - start_ts, 0)
```
Comment (isp-uetsuki) [inline src/caret_analyze/record/records_service/latency.py:100]: I suggest using variable defined above.

```suggestion
                    'latency': latency
```

File: src/caret_analyze/record/records_service/response_time.py [hunk 5]
```diff
@@ -16,13 +16,16 @@
...
 
+from logging import getLogger
 from warnings import warn
 
 from ..column import ColumnValue
 from ..interface import RecordInterface, RecordsInterface
 from ..record_factory import RecordFactory, RecordsFactory
 from ...common import ClockConverter
 
+logger = getLogger(__name__)
```
Comment (isp-uetsuki) [inline src/caret_analyze/record/records_service/response_time.py:27]: Is this change necessary? It seems `logger` is not used anywhere.

File: src/caret_analyze/record/records_service/response_time.py [hunk 6]
```diff
@@ -299,13 +308,14 @@ def to_worst_with_external_latency_case_records(
                                                          key=lambda x: x[0]):
             if converter:
                 record = {
-                    self._start_column: round(converter.convert(start_ts - worst_to_best_ts)),
-                    'response_time': end_ts - (start_ts - worst_to_best_ts)
+                    self._start_column: round(
+                        converter.convert(max(start_ts - worst_to_best_ts, 0))),
+                    'response_time': max(end_ts - (start_ts - worst_to_best_ts), 0)
```
Comment (isp-uetsuki) [inline src/caret_analyze/record/records_service/response_time.py:313]: From my understanding, we need to print warning and ignore data when its response time is lesser than 0. Can you modify this change and another changes in this file according to this policy?

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

File: src/caret_analyze/infra/lttng/lttng.py [hunk 7]
```diff
@@ -258,6 +258,42 @@ def _to_dicts(trace_dir: str, count: int) -> list[dict]:
...
+        self._next_id = 1000000000
+
+    def remap(self, event: dict):
+        target_id = get_field(event, self._id_key)
+        if target_id in self._other_host_ids or target_id in self._current_host_remapped_ids:
+            if target_id in self._current_host_id_map:
+                event[self._id_key] = self._current_host_id_map[target_id]
+            else:
+                while self._next_id in self._other_host_ids or \
+                        self._next_id in self._current_host_remapped_ids:
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/lttng.py]: Fixed in [6f491d8](https://github.com/tier4/caret_analyze/pull/455/commits/6f491d84507b655f6342ff642a4f5274859180d7).

File: src/caret_analyze/infra/lttng/lttng.py [hunk 8]
```diff
@@ -358,83 +398,105 @@ def _parse_lttng_data(
...
+        first_element_type = type(trace_dir_or_events[0])
+        if len(trace_dir_or_events) != len(Util.filter_items(
+                lambda x: isinstance(x, first_element_type), trace_dir_or_events)):
+            raise InvalidArgumentError(f'all elements of {trace_dir_or_events} \
+                                         must be same type: {first_element_type}.')
 
+        # TODO(hsgwa): Same implementation duplicated. Refactoring required.
+        if isinstance(trace_dir_or_events[0], str):
+            tid_remapper = MultiHostIdRemapper('_vtid')
+            pid_remapper = MultiHostIdRemapper('_vpid')
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/lttng.py]: Fixed in [f745ebe](https://github.com/tier4/caret_analyze/pull/455/commits/f745ebeb042a989db5747250c9779b815d6795f9).

File: src/caret_analyze/plot/timeseries/frequency_timeseries.py (hunk 1 above)
Comment (ymski) [reply src/caret_analyze/plot/timeseries/frequency_timeseries.py]: Fixed in [b78e69e](https://github.com/tier4/caret_analyze/pull/455/commits/b78e69ec501fdf07ea57a217b4b8931a4b23088f).

File: src/caret_analyze/record/records_service/latency.py (hunk 3 above)
Comment (ymski) [reply src/caret_analyze/record/records_service/latency.py]: Fixed in [cc43ec7](https://github.com/tier4/caret_analyze/pull/455/commits/cc43ec754a8a2f0cc4fc2f14d9479bde3042e19b).

File: src/caret_analyze/record/records_service/response_time.py (hunk 6 above)
Comment (ymski) [reply src/caret_analyze/record/records_service/response_time.py]: Fixed in [cc43ec7](https://github.com/tier4/caret_analyze/pull/455/commits/cc43ec754a8a2f0cc4fc2f14d9479bde3042e19b).

File: src/caret_analyze/record/records_service/latency.py (hunk 2 above)
Comment (ymski) [reply src/caret_analyze/record/records_service/latency.py]: Fixed in [cc43ec7](https://github.com/tier4/caret_analyze/pull/455/commits/cc43ec754a8a2f0cc4fc2f14d9479bde3042e19b).

File: src/caret_analyze/record/records_service/latency.py (hunk 4 above)
Comment (ymski) [reply src/caret_analyze/record/records_service/latency.py]: Fixed in [cc43ec7](https://github.com/tier4/caret_analyze/pull/455/commits/cc43ec754a8a2f0cc4fc2f14d9479bde3042e19b).

File: src/caret_analyze/record/records_service/response_time.py (hunk 5 above)
Comment (ymski) [reply src/caret_analyze/record/records_service/response_time.py]: Fixed in [6216774](https://github.com/tier4/caret_analyze/pull/455/commits/62167743fa765e4a6acda68e957041f46e6ff371).

File: src/caret_analyze/infra/lttng/lttng.py [hunk 9]
```diff
@@ -358,83 +398,105 @@ def _parse_lttng_data(
...
+            for trace_dir_or_event in trace_dir_or_events:
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/lttng.py]: Fixed in [34df5b1](https://github.com/tier4/caret_analyze/pull/455/commits/34df5b11e816f0cefdfa41ed9ed67b5c9ce99203).

File: src/caret_analyze/infra/lttng/lttng.py [hunk 10]
```diff
@@ -358,83 +398,105 @@ def _parse_lttng_data(
...
+
+                # Offset is obtained for conversion from
+                # the monotonic clock time to the system time.
+                for event in event_collection:
+                    event_name = event[LttngEventFilter.NAME]
+                    if event_name == 'ros2_caret:caret_init':
+                        offset = Ros2Handler.get_monotonic_to_system_offset(event)
+                        break
+
+                handler = Ros2Handler(data, offset)
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/caret_analyze/infra/lttng/lttng.py]: Fixed in [20557bd](https://github.com/tier4/caret_analyze/pull/455/commits/20557bd51925c7eac9bea929a0ac50c0feabcd27).

File: src/caret_analyze/infra/lttng/id_remapper.py [hunk 11]
```diff
@@ -101,3 +101,108 @@ def get_object_id(
             return max_item.remapped_id
         else:
             return addr
+
+
+class AddressRemapper:
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/id_remapper.py:106]: I suggest changing name of this class to `IDRemapperCollection` to better represent it's role.
Variable name `_remapper` looks good to me even after this change.

File: src/caret_analyze/infra/lttng/lttng.py [hunk 12]
```diff
@@ -358,83 +400,106 @@ def _parse_lttng_data(
...
+        if len(trace_dir_or_events) != len(Util.filter_items(
+                lambda x: isinstance(x, first_element_type), trace_dir_or_events)):
+            raise InvalidArgumentError(f'all elements of {trace_dir_or_events} \
+                                         must be same type: {first_element_type}.')
 
+        # TODO(hsgwa): Same implementation duplicated. Refactoring required.
+        if isinstance(trace_dir_or_events[0], str):
+            tid_remapper = MultiHostIdRemapper(LttngEventFilter.VTID)
+            pid_remapper = MultiHostIdRemapper(LttngEventFilter.VPID)
+            evant_remapper = AddressRemapper()
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/lttng.py:420]: Minor fix of typo.
```suggestion
            event_remapper = AddressRemapper()
```

File: src/caret_analyze/infra/lttng/ros2_tracing/processor.py [hunk 13]
```diff
@@ -64,34 +67,36 @@ def __init__(
...
+        # self._client_handle_remapper = IDRemapper()
+        # self._rmw_client_handle_remapper = IDRemapper()
+        # self._timer_handle_remapper = IDRemapper()
+        # self._state_machine_remapper = IDRemapper()
+        # self._executor_addr_remapper = IDRemapper()
+        # self._entities_collector_addr_remapper = IDRemapper()
+        # self._callback_group_addr_remapper = IDRemapper()
+        # self._callback_remapper = IDRemapper()
+        # self._buffer_remapper = IDRemapper()
+        # self._ipb_remapper = IDRemapper()
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/ros2_tracing/processor.py:98]: I suggest removing these comments.
```suggestion
```

File: src/caret_analyze/record/records_service/response_time.py (hunk 6 above)
Comment (isp-uetsuki) [reply src/caret_analyze/record/records_service/response_time.py]: I think current code does not follow the policy I mentioned above. (There is no value checking and warning printing.)
Can you check the code again?

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

File: src/caret_analyze/infra/lttng/lttng.py (hunk 12 above)
Comment (ymski) [reply src/caret_analyze/infra/lttng/lttng.py]: Fixed in [c67c785](https://github.com/tier4/caret_analyze/pull/455/commits/c67c7853ecf19972daffb03d6c53bbd6444449a8).

File: src/caret_analyze/infra/lttng/ros2_tracing/processor.py (hunk 13 above)
Comment (ymski) [reply src/caret_analyze/infra/lttng/ros2_tracing/processor.py]: Fixed in [1d9ae6e](https://github.com/tier4/caret_analyze/pull/455/commits/1d9ae6e9cab36add6e776e71f2a5443db1d64d4f).

File: src/caret_analyze/infra/lttng/id_remapper.py (hunk 11 above)
Comment (ymski) [reply src/caret_analyze/infra/lttng/id_remapper.py]: Fixed in [1d9ae6e](https://github.com/tier4/caret_analyze/pull/455/commits/1d9ae6e9cab36add6e776e71f2a5443db1d64d4f).

File: src/test/record/records_service/test_response_time_all.py [hunk 14]
```diff
@@ -1054,3 +1054,20 @@ def test_drop_case(self):
         ]
         result = to_dict(response_time.to_worst_case_stacked_bar())
         assert result == expect_raw
+
+    def test_invalid_value_case(self):
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/test/record/records_service/test_response_time_all.py:1074]: Fixed in [a672dc5](https://github.com/tier4/caret_analyze/pull/455/commits/a672dc540d2381f9e1ffefc68ccf4e7121c282fe).

File: src/test/infra/lttng/test_lttng.py [hunk 15]
```diff
@@ -1887,3 +1912,175 @@ def test_invalid_cache_exists_case(
...
+        host1_events = [
+            {'_vtid': 1000},
+            {'_vtid': 1001},
+            {'_vtid': 1002},
+        ]
+
+        host2_events = [
+            {'_vtid': 1003},
+            {'_vtid': 1001},  # duplicated id
+            {'_vtid': 1002},  # duplicated id
```
[the remark this answers is not in the document]
Comment (ymski) [reply src/test/infra/lttng/test_lttng.py]: I agree with your suggestion.
I have added the test in [a672dc5](https://github.com/tier4/caret_analyze/pull/455/commits/a672dc540d2381f9e1ffefc68ccf4e7121c282fe).

File: src/caret_analyze/infra/lttng/ros2_tracing/processor.py [hunk 16]
```diff
@@ -64,34 +67,16 @@ def __init__(
         ----------
         data : Ros2DataModel
             DataModel to be handles
+        remapper : IDRemapperCollection
+            Collections of IDRemapper.
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/ros2_tracing/processor.py:71]: Minor fix of comment.
```suggestion
            Collection of IDRemappers.
```

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

**Comment by isp-uetsuki** [comment]
LGTM

File: src/caret_analyze/infra/lttng/lttng.py [hunk 17]
```diff
@@ -370,79 +412,102 @@ def _parse_lttng_data(
...
+                        iter(run_events),
+                        total=len(run_events),
+                        desc='loading',
+                        mininterval=1.0):
+                    if store_events:
+                        event_dict = {
+                            k: get_field(event, k) for k in event
+                        }
+                        events.append(event_dict)
+                    tid_remapper.remap(event)
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/lttng.py:496]: Please revert `filtered_event_count += 1`. It is necessary to count filtered runtime events.

```suggestion
                        events.append(event_dict)
                    filtered_event_count += 1
                    tid_remapper.remap(event)
```

File: src/caret_analyze/infra/lttng/ros2_tracing/processor.py [hunk 18]
```diff
@@ -987,9 +999,9 @@ def _handle_callback_group_add_subscription(
         callback_group_addr = get_field(event, 'callback_group_addr')
         subscription_handle = get_field(event, 'subscription_handle')
 
-        callback_group_addr = self._callback_group_addr_remapper.get_nearest_object_id(
+        callback_group_addr = self._remapper.callback_group_addr_remapper.get_nearest_object_id(
             callback_group_addr, event)
-        subscription_handle = self._subscription_handle_remapper.get_nearest_object_id(
+        subscription_handle = self._remapper._subscription_handle_remapper.get_nearest_object_id(
```
Comment (isp-uetsuki) [inline src/caret_analyze/infra/lttng/ros2_tracing/processor.py:1004]: Fixing direct access to private member.
```suggestion
        subscription_handle = self._remapper.subscription_handle_remapper.get_nearest_object_id(
```

**Review by isp-uetsuki — CHANGES_REQUESTED** [review]

File: src/caret_analyze/infra/lttng/lttng.py (hunk 17 above)
Comment (isp-uetsuki) [reply src/caret_analyze/infra/lttng/lttng.py:496]: Sorry, I have been confusing. Current code is correct.

**Review by isp-uetsuki — DISMISSED** [review]
LGTM

**Review by xygyo77 — APPROVED** [review]
LGTM
